File: CPP/start_with_simple_nodes/src/node_timer_with_class.cpp
/**
 * @file node_timer_with_class.cpp
 *
 * @brief A basic ROS2 node implemented with class that spins
 *        and printouts a string based on a timer
 *
 * @author Antonio Mauro Galiano
 * Contact: https://www.linkedin.com/in/antoniomaurogaliano/
 *
 */


#include "rclcpp/rclcpp.hpp"


class MyNode : public rclcpp::Node
{
private:
  rclcpp::TimerBase::SharedPtr timer_;    // define a ROS2 timer object contained inside a shared ptr
  void TimerCallback();   // callback for the timer, what it will do when called

public:
  MyNode() : Node("ros2_node_class_timer")
  {
    // the timer is initialized with an inherit function from rclcpp::Node
    // passing two arguments: the frequency to call the callback and
    // the binded callback itself (binded to be able to pass the class method)
    // which is triggered when the node starts spinning
    timer_ = this->create_wall_timer(
            std::chrono::milliseconds(200),
            std::bind(&MyNode::TimerCallback,this));
  }
};


void MyNode::TimerCallback()    // simple callback triggered by the timer to printout info
{
  if (rclcpp::ok()) {
    RCLCPP_INFO_STREAM(this->get_logger(), "Echo line from a ROS2 class node with timer");
  }
}


int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<MyNode>();
  
  try {
    rclcpp::spin(node);
  } catch (const std::exception& e) {
    RCLCPP_ERROR(node->get_logger(), "Exception in spin: %s", e.what());
  }
  
  // Graceful shutdown: node will clean up timer automatically via RAII
  node.reset();
  rclcpp::shutdown();
  return 0;
}

File: PYTHON/start_with_simple_nodes_py/start_with_simple_nodes_py/node_timer_with_class.py
#  *
#  * @file node_timer_with_class.py
#  *
#  * @brief A basic ROS2 node implemented with class that spins
#  *        and printouts a string based on a timer
#  *
#  * @author Antonio Mauro Galiano
#  * Contact: https://www.linkedin.com/in/antoniomaurogaliano/
#  *
#  *


import rclpy
from rclpy.node import Node

class MyFirstNode(Node):
    def __init__(self):
        super().__init__('my_first_node_timer_class')
        self.create_timer(0.2, self.timer_callback)

    def timer_callback(self):
        self.get_logger().info("Echo line from a ROS2 class timer callback ")

def main(args=None):
    rclpy.init(args=args)
    node = MyFirstNode()
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == '__main__':
    main()